bulk_insert.go

```diff
@@ -25,79 +25,58 @@ import (
+//
 // [chunkSize] is a number of variables embedded in query. To prevent the error which occurs embedding a large number of variables at once
 // and exceeds the limit of prepared statement. Larger size normally leads to better performance, in most cases 2000 to 3000 is reasonable.
 //
 // [excludeColumns] is column names to exclude from insert.
-func BulkInsertWithAssigningIDs(db *gorm.DB, returnedValue interface{}, objects []interface{}, chunkSize int, excludeColumns ...string) error {
-	typ := reflect.TypeOf(returnedValue)
-	if typ.Kind() != reflect.Ptr || typ.Elem().Kind() != reflect.Slice {
-		return errors.New("returningId must be a slice ptr")
+func BulkInsertWithReturningValues(db *gorm.DB, objects []interface{}, returnedVals interface{}, chunkSize int, excludeColumns ...string) error {
+	typ := reflect.TypeOf(returnedVals)
+	if typ.Kind() != reflect.Ptr || typ.Elem().Kind() != reflect.Slice || typ.Elem().Elem().Kind() != reflect.Struct {
+		return errors.New("returnedVals must be a pointer to a slice of struct")
 	}
 
-	allIds := reflect.Indirect(reflect.ValueOf(returnedValue))
-	typ = allIds.Type()
-
-	// Deference value of slice
-	valueTyp := typ.Elem()
-	for valueTyp.Kind() == reflect.Ptr {
-		valueTyp = valueTyp.Elem()
-	}
+	refDst := reflect.Indirect(reflect.ValueOf(returnedVals))
+	typ = refDst.Type()
 
 	// Split records with specified size not to exceed Database parameter limit
 	for _, objSet := range splitObjects(objects, chunkSize) {
-		returnValueSlice := reflect.New(typ)
-		var scanReturningId func(*gorm.DB) error
-		switch valueTyp.Kind() {
-		case reflect.Struct:
-			// If user want to scan `returning *` with returnedValue=[]struct{...}
-			scanReturningId = func(db *gorm.DB) error {
-				return db.Scan(returnValueSlice.Interface()).Error
-			}
-		default:
-			// If user want to scan primary key `returning pk` with returnedValue=[]struct{...}
-			pk := db.NewScope(objects[0]).PrimaryKey()
-			scanReturningId = func(db *gorm.DB) error {
-				return db.Pluck(pk, returnValueSlice.Interface()).Error
-			}
+		db, err := insertObjSet(db, objSet, excludeColumns...)
+		if err != nil {
+			return err
 		}
-
-		if err := insertObjSetWithCallback(db, objSet, scanReturningId, excludeColumns...); err != nil {
+		scanned := reflect.New(typ)
+		if err := db.Scan(scanned.Interface()).Error; err != nil {
 			return err
 		}
-
-		allIds.Set(reflect.AppendSlice(allIds, returnValueSlice.Elem()))
+		refDst.Set(reflect.AppendSlice(refDst, scanned.Elem()))
 	}
 	return nil
 }
 
-func insertObjSet(db *gorm.DB, objects []interface{}, excludeColumns ...string) error {
-	return insertObjSetWithCallback(db, objects, nil, excludeColumns...)
-}
-
-func insertObjSetWithCallback(db *gorm.DB, objects []interface{}, postInsert func(*gorm.DB) error, excludeColumns ...string) error {
+func insertObjSet(db *gorm.DB, objects []interface{}, excludeColumns ...string) (*gorm.DB, error) {
```

Comment:
Modified the private method of `insertObjSet` to return both `gorm.DB` and `error` rather receiving the optional `postInsert` function.

Reply:
😉how about append err to db then return single `*gorm.DB` like gorm?

Reply:
I don't think that's a good idea because I think we should be able to tell the difference between Gorm internal errors added to the `gorm.DB` and errors coming from `gorm-bulk-insert`. 

I would rather keep the old interface just returning the `error` since we pass a pointer to the `gorm.DB`, any modification to it will be readable from the caller.

Reply:
Or maybe you're right, doesn't feel nice to have to check error twice so maybe best to add them to `gorm.DB`? 🤔 Either or works for me 😸

Reply:
> I would rather keep the old interface just returning the error since we pass a pointer to the gorm.DB, any modification to it will be readable from the caller.

From my understanding, we need to return `*gorm.DB` unless we use reflection. The reason is as follows.

- `db.Raw` returns a new pointer to DB and passed "db" is not affected.
- If we re-assign "db" in line 139, it does not affect to the caller since it just rewrites the pointer. Pointer is just a value that holds the memory address, therefore, re-assign does not affect the caller.
- Most of the fields of `gorm.DB` are private, so we cannot copy the fields into "db" of argument. So, I think we have to return "db" after executing "db.Raw(xx)".

Please correct me if I'm wrong🙇‍♂️

Reply:
I guess you can dereference and change what the reference points to, example: https://play.golang.org/p/DO2Tpezj9hq

Reply:
Oh, I did not know such a basic thing😅 Thank you, I'll revert the return values.